File: tools/tf-types-ts/src/core/policy-engine.ts
/**
 * Native TrustForge policy engine.
 *
 * Loads a `policy.schema.json` manifest and answers a `PolicyQuery` with
 * a `PolicyDecision`. The manifest is the authoritative input; rules are
 * evaluated top-to-bottom, with negative_capabilities applied first
 * (overriding any allow downstream) and `continuous_reevaluation` triggers
 * surfaced for callers (typically the RpcServer.reevaluate hook).
 *
 * The engine is intentionally pluggable: callers can swap in a Cedar or
 * Rego adapter by providing a `PolicyEngine` implementation. v0.1.0 ships
 * the native engine and adapter stubs that throw "not implemented" until
 * those backends are wired up.
 */

import type { Policy } from "../generated/policy.js";
import type { PolicyDecision } from "../generated/policy-decision.js";
import type {
  ActorId,
  ActionName,
  Capability,
  Constraint,
  EnforcementLevel,
  NegativeCapability,
  ProofLevel,
  Timestamp,
} from "../generated/_common.js";
import { sha256 } from "@noble/hashes/sha256";
import { canonicalize } from "./canonical.js";

export interface PolicyQuery {
  subject: ActorId;
  instance?: string;
  action: ActionName;
  target?: string;
  context?: Record<string, unknown>;
  /** Active negative capabilities for this evaluation. Engines apply
   *  these BEFORE walking the rules. */
  negativeCapabilities?: NegativeCapability[];
  /** Active enforcement level. Forwarded into the decision so audit
   *  logs preserve the runtime posture. */
  enforcementLevel?: EnforcementLevel;
  /** Wall-clock timestamp of the evaluation in RFC 3339; defaults to
   *  `new Date().toISOString()`. */
  now?: Timestamp;
}

export interface PolicyEngine {
  readonly engine: "native" | "cedar" | "rego" | "custom" | "none";
  evaluate(query: PolicyQuery): PolicyDecision;
}

export interface NativePolicyEngineOptions {
  policy: Policy;
  /** Optional capability bag the engine consults to ensure the subject
   *  even has authority to attempt the action. When unset, rules are
   *  the only authority surface. */
  grantedCapabilities?: Capability[];
}

export class NativePolicyEngine implements PolicyEngine {
  readonly engine = "native" as const;
  private readonly policy: Policy;
  private readonly manifestHash: string;
  private readonly granted: Capability[];

  constructor(opts: NativePolicyEngineOptions) {
    this.policy = opts.policy;
    this.manifestHash = `sha256-${toHex(sha256(canonicalize(this.policy as unknown)))}`;
    this.granted = opts.grantedCapabilities ?? [];
  }

  evaluate(query: PolicyQuery): PolicyDecision {
    const now = query.now ?? new Date().toISOString();
    const negativeCaps =
      query.negativeCapabilities ??
      ((this.policy as unknown as Record<string, unknown>)["negative_capabilities"] as
        | NegativeCapability[]
        | undefined) ??
      [];

    // 1. Negative caps win.
    for (const neg of negativeCaps) {
      if (negativeCapMatches(neg, query)) {
        return this.decision({
          query,
          decision: "deny",
          reason: neg.reason || `denied by negative_capability ${neg.name}`,
          ruleId: undefined,
          now,
          negCaps: negativeCaps,
        });
      }
    }

    // 2. Walk the rules in order; first match wins.
    for (const rule of this.policy.rules) {
      if (!ruleMatches(rule, query)) continue;
      const constraints = (rule as unknown as { constraints?: Constraint[] }).constraints ?? [];
      const reason =
        (rule as unknown as { reason?: string }).reason ?? `matched rule ${rule.id}`;
      const proof = (rule as unknown as { proof_required?: ProofLevel }).proof_required;
      const approval = (rule as unknown as { approval?: PolicyDecision["approval"] }).approval;
      switch (rule.effect) {
        case "allow":
          return this.decision({
            query,
            decision: "allow",
            reason,
            ruleId: rule.id,
            constraints,
            proof,
            approval,
            now,
            negCaps: negativeCaps,
          });
        case "deny":
          return this.decision({
            query,
            decision: "deny",
            reason,
            ruleId: rule.id,
            now,
            negCaps: negativeCaps,
          });
        case "escalate":
          return this.decision({
            query,
            decision: approval === "quorum" ? "escalate" : "approval-required",
            reason,
            ruleId: rule.id,
            constraints,
            proof,
            approval: approval ?? "required",
            now,
            negCaps: negativeCaps,
          });
        case "log_only":
          return this.decision({
            query,
            decision: "log-only",
            reason,
            ruleId: rule.id,
            constraints,
            proof,
            now,
            negCaps: negativeCaps,
          });
      }
    }

    // 3. Default deny — the policy did not match any rule.
    return this.decision({
      query,
      decision: "deny",
      reason: "no matching rule (default deny)",
      now,
      negCaps: negativeCaps,
    });
  }

  /** Triggers the policy declares for in-flight reevaluation. Callers
   *  forward these to the RpcServer's `continuousReevaluation` config. */
  continuousTriggers(): string[] {
    const cont = (this.policy as unknown as Record<string, unknown>)["continuous_reevaluation"];
    if (!cont || typeof cont !== "object") return [];
    const triggers = (cont as { triggers?: string[] }).triggers ?? [];
    return triggers.slice();
  }

  /** Quorum defaults from the manifest, if any. */
  quorumDefaults(): { min_approvers: number; of: ActorId[] } | undefined {
    const q = (this.policy as unknown as Record<string, unknown>)["quorum_defaults"];
    if (!q || typeof q !== "object") return undefined;
    return q as { min_approvers: number; of: ActorId[] };
  }

  /** Capabilities the engine was constructed with. Useful for callers
   *  that need to enforce capability-bearer checks alongside the rules. */
  grantedCapabilities(): Capability[] {
    return this.granted.slice();
  }

  private decision(params: {
    query: PolicyQuery;
    decision: PolicyDecision["decision"];
    reason: string;
    ruleId?: string;
    constraints?: Constraint[];
    proof?: ProofLevel;
    approval?: PolicyDecision["approval"];
    now: Timestamp;
    negCaps: NegativeCapability[];
  }): PolicyDecision {
    const out: PolicyDecision = {
      decision_version: "1",
      policy_engine: "native",
      engine_version: "tf-policy-native-0.1.0",
      trust_domain: this.policy.trust_domain,
      subject: params.query.subject,
      action: params.query.action,
      decision: params.decision,
      evaluated_at: params.now,
      policy_manifest_hash: this.manifestHash,
    };
    if (params.query.instance) out.instance = params.query.instance;
    if (params.query.target) out.target = params.query.target;
    if (params.ruleId) out.rule_id = params.ruleId;
    if (params.reason) out.reason = params.reason;
    if (params.constraints && params.constraints.length > 0) out.constraints_applied = params.constraints;
    if (params.proof) out.proof_required = params.proof;
    if (params.approval) out.approval = params.approval;
    if (params.query.enforcementLevel) out.enforcement_level = params.query.enforcementLevel;
    if (params.negCaps.length > 0) out.negative_capabilities_consulted = params.negCaps;
    if (params.query.context && Object.keys(params.query.context).length > 0) out.context = params.query.context;
    return out;
  }
}

function ruleMatches(
  rule: Policy["rules"][number],
  query: PolicyQuery,
): boolean {
  if ((rule as unknown as { action?: string }).action) {
    if ((rule as unknown as { action: string }).action !== query.action) return false;
  }
  const pattern = (rule as unknown as { action_pattern?: string }).action_pattern;
  if (pattern) {
    // Match through globMatch instead of `new RegExp` to:
    //   1. avoid catastrophic-backtracking ReDoS from untrusted policies
    //   2. give TS and Rust identical match semantics (both implement
    //      the same minimal `*` / `**` glob).
    if (!globMatch(pattern, query.action)) return false;
  }
  const subjectPattern = (rule as unknown as { subject_pattern?: string }).subject_pattern;
  if (subjectPattern) {
    if (!globMatch(subjectPattern, query.subject)) return false;
  }
  const targets = (rule as unknown as { target_patterns?: string[] }).target_patterns ?? [];
  if (targets.length > 0) {
    if (!query.target) return false;
    const hit = targets.some((p) => globMatch(p, query.target!));
    if (!hit) return false;
  }
  return true;
}

function negativeCapMatches(neg: NegativeCapability, q: PolicyQuery): boolean {
  // Negative-capability `name` is glob-matched against the action so
  // patterns like `fs.write*` cover `fs.write.tmp`. Pre-B8 the match
  // was `===`, which only blocked exact action names.
  if (!globMatch(neg.name, q.action)) return false;
  if (!neg.target) return true;
  if (!q.target) return false;
  return globMatch(neg.target, q.target);
}

function globMatch(pattern: string, value: string): boolean {
  let re = "^";
  for (let i = 0; i < pattern.length; i++) {
    const c = pattern[i]!;
    if (c === "*") {
      if (pattern[i + 1] === "*") {
        re += ".*";
        i += 1;
      } else {
        re += "[^/]*";
      }
    } else if (".+^${}()|[]\\?".includes(c)) {
      re += "\\" + c;
    } else {
      re += c;
    }
  }
  return new RegExp(re + "$").test(value);
}

function toHex(bytes: Uint8Array): string {
  let out = "";
  for (const b of bytes) out += b.toString(16).padStart(2, "0");
  return out;
}

/**
 * Cedar adapter stub. v0.1.0 ships the indirection so callers don't
 * have to refactor when the real Cedar runtime lands; until then the
 * engine returns a graceful deny so a misconfigured `engine_hint:
 * cedar` doesn't crash the daemon mid-RPC. Pre-B8 this threw, which
 * caused the daemon's enforcer to surface "internal: capability
 * enforcer threw" errors to clients.
 */
function unavailableDecision(engine: "cedar" | "rego", q: PolicyQuery): PolicyDecision {
  const out: PolicyDecision = {
    decision_version: "1",
    policy_engine: engine,
    engine_version: `${engine}-stub-0.1.0`,
    trust_domain: "unknown",
    subject: q.subject,
    action: q.action,
    decision: "deny",
    evaluated_at: new Date().toISOString(),
    reason: `${engine} adapter not implemented in v0.1.0; configure engine_hint: native or wait for the v0.2.0 ${engine}-wasm integration`,
  };
  if (q.target) out.target = q.target;
  if (q.enforcementLevel) out.enforcement_level = q.enforcementLevel;
  return out;
}

export class CedarPolicyEngine implements PolicyEngine {
  readonly engine = "cedar" as const;
  evaluate(q: PolicyQuery): PolicyDecision {
    return unavailableDecision("cedar", q);
  }
}

/** Rego adapter — same graceful-deny shape as Cedar's. */
export class RegoPolicyEngine implements PolicyEngine {
  readonly engine = "rego" as const;
  evaluate(q: PolicyQuery): PolicyDecision {
    return unavailableDecision("rego", q);
  }
}

/** Build the right engine for a manifest's `engine_hint`. */
export function policyEngineForManifest(policy: Policy, opts: { grantedCapabilities?: Capability[] } = {}): PolicyEngine {
  const hint = (policy as unknown as { engine_hint?: string }).engine_hint;
  switch (hint) {
    case "cedar":
      return new CedarPolicyEngine();
    case "rego":
      return new RegoPolicyEngine();
    case "native":
    case "custom":
    case "none":
    case undefined:
      return new NativePolicyEngine({ policy, grantedCapabilities: opts.grantedCapabilities });
    default:
      throw new Error(`unsupported engine_hint: ${hint}`);
  }
}

File: crates/tf-types/src/policy_engine.rs
//! Native TrustForge policy engine — Rust mirror of
//! `tools/tf-types-ts/src/core/policy-engine.ts`.

use std::collections::HashMap;

use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};

use crate::canonicalize;
use crate::guard::NegativeCapability;

#[derive(Clone, Debug, Serialize, Deserialize, Default)]
pub struct PolicyQuery {
    pub subject: String,
    #[serde(default)]
    pub instance: Option<String>,
    pub action: String,
    #[serde(default)]
    pub target: Option<String>,
    #[serde(default)]
    pub context: HashMap<String, Value>,
    #[serde(default)]
    pub negative_capabilities: Vec<NegativeCapability>,
    #[serde(default)]
    pub enforcement_level: Option<String>,
    #[serde(default)]
    pub now: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct PolicyDecision {
    pub decision_version: String,
    pub policy_engine: String,
    pub engine_version: Option<String>,
    pub trust_domain: String,
    pub subject: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub instance: Option<String>,
    pub action: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub target: Option<String>,
    pub decision: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub rule_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub approval: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub proof_required: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub constraints_applied: Option<Vec<Value>>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub negative_capabilities_consulted: Option<Vec<NegativeCapability>>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub enforcement_level: Option<String>,
    pub evaluated_at: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub policy_manifest_hash: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub context: Option<HashMap<String, Value>>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PolicyRule {
    pub id: String,
    pub effect: String,
    #[serde(default)]
    pub action: Option<String>,
    #[serde(default)]
    pub action_pattern: Option<String>,
    #[serde(default)]
    pub subject_pattern: Option<String>,
    #[serde(default)]
    pub target_patterns: Option<Vec<String>>,
    #[serde(default)]
    pub approval: Option<String>,
    #[serde(default)]
    pub proof_required: Option<String>,
    #[serde(default)]
    pub constraints: Option<Vec<Value>>,
    #[serde(default)]
    pub reason: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PolicyManifest {
    pub policy_version: String,
    pub trust_domain: String,
    #[serde(default)]
    pub engine_hint: Option<String>,
    pub rules: Vec<PolicyRule>,
    #[serde(default)]
    pub negative_capabilities: Vec<NegativeCapability>,
    #[serde(default)]
    pub continuous_reevaluation: Option<ContinuousReeval>,
    #[serde(default)]
    pub quorum_defaults: Option<QuorumDefaults>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ContinuousReeval {
    pub triggers: Vec<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct QuorumDefaults {
    pub min_approvers: u32,
    pub of: Vec<String>,
}

/// A pluggable policy engine. Implemented by the native engine, the
/// `tf-cedar` crate, and the `tf-rego` crate. Decoupling via a trait
/// (rather than a feature-gated dependency on the adapter crates) lets
/// `tf-types` stay lightweight while still letting the daemon dispatch
/// the right engine for a given `engine_hint`.
pub trait PolicyEngineImpl {
    fn evaluate(&self, query: &PolicyQuery) -> PolicyDecision;
}

pub struct NativePolicyEngine {
    policy: PolicyManifest,
    manifest_hash: String,
}

impl PolicyEngineImpl for NativePolicyEngine {
    fn evaluate(&self, query: &PolicyQuery) -> PolicyDecision {
        NativePolicyEngine::evaluate(self, query)
    }
}

/// Dispatch a `PolicyQuery` to the appropriate backend based on
/// `engine_hint`. Pass an explicit backend for the hints that need one;
/// `native` falls back to the supplied native engine. When the requested
/// hint has no backend wired in (e.g. caller didn't construct a Cedar
/// engine yet) the dispatcher returns a safe deny.
///
/// The signature uses `dyn` trait objects so callers don't have to leak
/// the cedar / rego crate types through `tf-types`. `tf-cedar` and
/// `tf-rego` each export an adapter that implements `PolicyEngineImpl`.
pub fn evaluate_with_engine(
    hint: Option<&str>,
    native: &NativePolicyEngine,
    cedar: Option<&dyn PolicyEngineImpl>,
    rego: Option<&dyn PolicyEngineImpl>,
    query: &PolicyQuery,
) -> PolicyDecision {
    match hint {
        Some("cedar") => match cedar {
            Some(eng) => eng.evaluate(query),
            None => unavailable_decision("cedar", query, native.policy.trust_domain.as_str()),
        },
        Some("rego") => match rego {
            Some(eng) => eng.evaluate(query),
            None => unavailable_decision("rego", query, native.policy.trust_domain.as_str()),
        },
        _ => native.evaluate(query),
    }
}

fn unavailable_decision(engine: &str, query: &PolicyQuery, trust_domain: &str) -> PolicyDecision {
    PolicyDecision {
        decision_version: "1".into(),
        policy_engine: engine.into(),
        engine_version: Some(format!("{engine}-stub")),
        trust_domain: trust_domain.into(),
        subject: query.subject.clone(),
        instance: query.instance.clone(),
        action: query.action.clone(),
        target: query.target.clone(),
        decision: "deny".into(),
        rule_id: None,
        reason: Some(format!(
            "{engine} engine not configured for this dispatcher (no adapter supplied)"
        )),
        approval: None,
        proof_required: None,
        constraints_applied: None,
        negative_capabilities_consulted: None,
        enforcement_level: query.enforcement_level.clone(),
        evaluated_at: now_iso8601(),
        policy_manifest_hash: None,
        context: if query.context.is_empty() {
            None
        } else {
            Some(query.context.clone())
        },
    }
}

impl NativePolicyEngine {
    pub fn new(policy: PolicyManifest) -> Self {
        let canonical_value = serde_json::to_value(&policy).unwrap_or(Value::Null);
        let canonical = canonicalize(&canonical_value).unwrap_or_default();
        let digest: [u8; 32] = Sha256::digest(canonical.as_bytes()).into();
        let hex: String = digest.iter().map(|b| format!("{:02x}", b)).collect();
        let manifest_hash = format!("sha256-{}", hex);
        NativePolicyEngine {
            policy,
            manifest_hash,
        }
    }

    pub fn evaluate(&self, query: &PolicyQuery) -> PolicyDecision {
        let now = query.now.clone().unwrap_or_else(|| now_iso8601());
        let neg_caps = if query.negative_capabilities.is_empty() {
            self.policy.negative_capabilities.clone()
        } else {
            query.negative_capabilities.clone()
        };
        for neg in &neg_caps {
            if negative_matches(neg, query) {
                return self.decision(
                    query,
                    "deny",
                    neg.reason
                        .clone()
                        .unwrap_or_else(|| format!("denied by negative_capability {}", neg.name)),
                    None,
                    None,
                    None,
                    None,
                    Some(&neg_caps),
                    &now,
                );
            }
        }
        for rule in &self.policy.rules {
            if !rule_matches(rule, query) {
                continue;
            }
            let reason = rule
                .reason
                .clone()
                .unwrap_or_else(|| format!("matched rule {}", rule.id));
            match rule.effect.as_str() {
                "allow" => {
                    return self.decision(
                        query,
                        "allow",
                        reason,
                        Some(rule.id.clone()),
                        rule.constraints.clone(),
                        rule.proof_required.clone(),
                        rule.approval.clone(),
                        Some(&neg_caps),
                        &now,
                    );
                }
                "deny" => {
                    return self.decision(
                        query,
                        "deny",
                        reason,
                        Some(rule.id.clone()),
                        None,
                        None,
                        None,
                        Some(&neg_caps),
                        &now,
                    );
                }
                "escalate" => {
                    let decision = if rule.approval.as_deref() == Some("quorum") {
                        "escalate"
                    } else {
                        "approval-required"
                    };
                    return self.decision(
                        query,
                        decision,
                        reason,
                        Some(rule.id.clone()),
                        rule.constraints.clone(),
                        rule.proof_required.clone(),
                        rule.approval.clone().or_else(|| Some("required".into())),
                        Some(&neg_caps),
                        &now,
                    );
                }
                "log_only" => {
                    return self.decision(
                        query,
                        "log-only",
                        reason,
                        Some(rule.id.clone()),
                        rule.constraints.clone(),
                        rule.proof_required.clone(),
                        None,
                        Some(&neg_caps),
                        &now,
                    );
                }
                _ => continue,
            }
        }
        self.decision(
            query,
            "deny",
            "no matching rule (default deny)".into(),
            None,
            None,
            None,
            None,
            Some(&neg_caps),
            &now,
        )
    }

    pub fn continuous_triggers(&self) -> Vec<String> {
        self.policy
            .continuous_reevaluation
            .as_ref()
            .map(|c| c.triggers.clone())
            .unwrap_or_default()
    }

    pub fn quorum_defaults(&self) -> Option<&QuorumDefaults> {
        self.policy.quorum_defaults.as_ref()
    }

    pub fn manifest_hash(&self) -> &str {
        &self.manifest_hash
    }

    #[allow(clippy::too_many_arguments)]
    fn decision(
        &self,
        query: &PolicyQuery,
        decision: &str,
        reason: String,
        rule_id: Option<String>,
        constraints: Option<Vec<Value>>,
        proof: Option<String>,
        approval: Option<String>,
        neg_caps: Option<&[NegativeCapability]>,
        now: &str,
    ) -> PolicyDecision {
        PolicyDecision {
            decision_version: "1".into(),
            policy_engine: "native".into(),
            engine_version: Some("tf-policy-native-0.1.0".into()),
            trust_domain: self.policy.trust_domain.clone(),
            subject: query.subject.clone(),
            instance: query.instance.clone(),
            action: query.action.clone(),
            target: query.target.clone(),
            decision: decision.into(),
            rule_id,
            reason: Some(reason),
            approval,
            proof_required: proof,
            constraints_applied: constraints.filter(|c| !c.is_empty()),
            negative_capabilities_consulted: neg_caps.map(|c| c.to_vec()).filter(|v| !v.is_empty()),
            enforcement_level: query.enforcement_level.clone(),
            evaluated_at: now.into(),
            policy_manifest_hash: Some(self.manifest_hash.clone()),
            context: if query.context.is_empty() {
                None
            } else {
                Some(query.context.clone())
            },
        }
    }
}

fn rule_matches(rule: &PolicyRule, query: &PolicyQuery) -> bool {
    if let Some(action) = &rule.action {
        if action != &query.action {
            return false;
        }
    }
    if let Some(pattern) = &rule.action_pattern {
        let re = match Regex::new(pattern) {
            Ok(r) => r,
            Err(_) => return false,
        };
        if !re.is_match(&query.action) {
            return false;
        }
    }
    if let Some(pattern) = &rule.subject_pattern {
        let re = match Regex::new(pattern) {
            Ok(r) => r,
            Err(_) => return false,
        };
        if !re.is_match(&query.subject) {
            return false;
        }
    }
    if let Some(targets) = &rule.target_patterns {
        if !targets.is_empty() {
            let Some(target) = &query.target else {
                return false;
            };
            if !targets.iter().any(|p| glob_match(p, target)) {
                return false;
            }
        }
    }
    true
}

fn negative_matches(neg: &NegativeCapability, q: &PolicyQuery) -> bool {
    if neg.name != q.action {
        return false;
    }
    let Some(target_pattern) = neg.target.as_deref() else {
        return true;
    };
    let Some(query_target) = q.target.as_deref() else {
        return false;
    };
    glob_match(target_pattern, query_target)
}

fn glob_match(pattern: &str, value: &str) -> bool {
    let mut re = String::from("^");
    let bytes = pattern.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        let b = bytes[i];
        match b {
            b'*' => {
                if i + 1 < bytes.len() && bytes[i + 1] == b'*' {
                    re.push_str(".*");
                    i += 2;
                } else {
                    re.push_str("[^/]*");
                    i += 1;
                }
            }
            b'.' | b'+' | b'^' | b'$' | b'{' | b'}' | b'(' | b')' | b'|' | b'[' | b']' | b'\\' => {
                re.push('\\');
                re.push(b as char);
                i += 1;
            }
            _ => {
                re.push(b as char);
                i += 1;
            }
        }
    }
    re.push('$');
    Regex::new(&re).map(|r| r.is_match(value)).unwrap_or(false)
}

fn now_iso8601() -> String {
    let secs = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs() as i64;
    let (year, month, day, hour, minute, second) = secs_to_ymdhms(secs);
    format!(
        "{:04}-{:02}-{:02}T{:02}:{:02}:{:02}Z",
        year, month, day, hour, minute, second
    )
}

fn secs_to_ymdhms(secs: i64) -> (i32, u32, u32, u32, u32, u32) {
    let days = secs.div_euclid(86_400);
    let time = secs.rem_euclid(86_400);
    let hour = (time / 3600) as u32;
    let minute = ((time % 3600) / 60) as u32;
    let second = (time % 60) as u32;
    let z = days + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = (z - era * 146_097) as u64;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe as i64 + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = (doy - (153 * mp + 2) / 5 + 1) as u32;
    let m = if mp < 10 {
        (mp + 3) as u32
    } else {
        (mp - 9) as u32
    };
    let year = if m <= 2 { y + 1 } else { y };
    (year as i32, m, d, hour, minute, second)
}
